### ISSUE TITLE
PHP xds tests break after upb changes

### ISSUE THREAD

**Issue by stanley-cheung** [grpc/grpc#23307]
I noticed that the PHP xds tests (https://fusion.corp.google.com/projectanalysis/summary/KOKORO/prod%3Agrpc%2Fcore%2Fmaster%2Flinux%2Fgrpc_xds_php) start failing after #21941 is merged.

Here's the error if I run the PHP extension build on `master` now:

```
 g++ -fvisibility=hidden -DOPENSSL_NO_ASM -D_GNU_SOURCE -DWIN32_LEAN_AND_MEAN -D_HAS_EXCEPTIONS=0 
-DNOMINMAX -DGRPC_ARES=0 -DGRPC_POSIX_FORK_ALLOW_PTHREAD_ATFORK=1 -I. -I/tmp/pear/temp/grpc 
-DPHP_ATOM_INC -I/tmp/pear/temp/pear-build-defaultuserW60hNa/grpc-1.31.0dev/include 
-I/tmp/pear/temp/pear-build-defaultuserW60hNa/grpc-1.31.0dev/main -I/tmp/pear/temp/grpc -I/usr/local/include/php 
-I/usr/local/include/php/main -I/usr/local/include/php/TSRM -I/usr/local/include/php/Zend -I/usr/local/include/php/ext 
-I/usr/local/include/php/ext/date/lib -I/tmp/pear/temp/grpc/include -I/tmp/pear/temp/grpc/src/core/ext/upb-generated 
-I/tmp/pear/temp/grpc/src/php/ext/grpc -I/tmp/pear/temp/grpc/third_party/abseil-cpp 
-I/tmp/pear/temp/grpc/third_party/address_sorting/include -I/tmp/pear/temp/grpc/third_party/boringssl-with-bazel/src/include 
-I/tmp/pear/temp/grpc/third_party/upb -DHAVE_CONFIG_H -std=c++11 -fno-exceptions -fno-rtti -g -O2 
-c /tmp/pear/temp/grpc/src/core/ext/filters/client_channel/lb_policy/xds/cds.cc  -fPIC -DPIC
-o src/core/ext/filters/client_channel/lb_policy/xds/.libs/cds.o
In file included from /tmp/pear/temp/grpc/src/core/ext/filters/client_channel/xds/xds_client.h:28:0,          
                 from /tmp/pear/temp/grpc/src/core/ext/filters/client_channel/lb_policy/xds/cds.cc:27:        
/tmp/pear/temp/grpc/src/core/ext/filters/client_channel/xds/xds_api.h:31:23: fatal error: upb/def.hpp: No such file or directory   
 #include "upb/def.hpp"                                                                                       
                       ^                                                                                      
compilation terminated.                                                                                       
Makefile:246: recipe for target 'src/core/ext/filters/client_channel/lb_policy/xds/cds.lo' failed             
make: *** [src/core/ext/filters/client_channel/lb_policy/xds/cds.lo] Error 1                                  
make: *** Waiting for unfinished jobs....                                                                     
ERROR: `make' failed                                  
```

I noticed that `def.hpp` is not part of `upb` [public headers](https://github.com/protocolbuffers/upb/blob/master/BUILD#L78), so it was never pulled into this PHP `package.xml` file (which is the source list for all the files PHP needs to build its extension): https://github.com/grpc/grpc/blob/master/package.xml#L1716-L1735.

So I tried to manually add that `def.hpp` file into my `package.xml` file, but then I got these errors now:

```
src/core/ext/upb-generated/envoy/annotations/.libs/deprecation.o:(.data.rel.local+0x0): multiple definition of `envoy_annotations_deprecation_proto_upbdefinit'
src/core/ext/upb-generated/envoy/annotations/.libs/deprecation.o:(.data.rel.local+0x0): first defined here
src/core/ext/upb-generated/envoy/annotations/.libs/resource.o:(.data.rel.local+0x0): multiple definition of `envoy_annotations_resource_proto_upbdefinit'
src/core/ext/upb-generated/envoy/annotations/.libs/resource.o:(.data.rel.local+0x0): first defined here
src/core/ext/upb-generated/envoy/api/v2/auth/.libs/cert.o:(.data.rel.local+0x0): multiple definition of `envoy_api_v2_auth_cert_proto_upbdefinit'
src/core/ext/upb-generated/envoy/api/v2/auth/.libs/cert.o:(.data.rel.local+0x0): first defined here
src/core/ext/upb-generated/envoy/api/v2/auth/.libs/common.o:(.data.rel.local+0x0): multiple definition of `envoy_api_v2_auth_common_proto_upbdefinit'
src/core/ext/upb-generated/envoy/api/v2/auth/.libs/common.o:(.data.rel.local+0x0): first defined here
src/core/ext/upb-generated/envoy/api/v2/auth/.libs/secret.o:(.data.rel.local+0x0): multiple definition of `envoy_api_v2_auth_secret_proto_upbdefinit'
src/core/ext/upb-generated/envoy/api/v2/auth/.libs/secret.o:(.data.rel.local+0x0): first defined here
src/core/ext/upb-generated/envoy/api/v2/auth/.libs/tls.o:(.data.rel.local+0x0): multiple definition of `envoy_api_v2_auth_tls_proto_upbdefinit'
src/core/ext/upb-generated/envoy/api/v2/auth/.libs/tls.o:(.data.rel.local+0x0): first defined here
src/core/ext/upb-generated/envoy/api/v2/.libs/cds.o:(.data.rel.local+0x0): multiple definition of `envoy_api_v2_cds_proto_upbdefinit'
src/core/ext/upb-generated/envoy/api/v2/.libs/cds.o:(.data.rel.local+0x0): first defined here
src/core/ext/upb-generated/envoy/api/v2/.libs/cluster.o:(.data.rel.local+0x0): multiple definition of `envoy_api_v2_cluster_proto_upbdefinit'
src/core/ext/upb-generated/envoy/api/v2/.libs/cluster.o:(.data.rel.local+0x0): first defined here
```
Full log here: https://g3c.corp.google.com/results/invocations/82203764-4856-4096-8b8c-b0e0edeba3b1/targets/grpc%2Fcore%2Fexperimental%2Flinux%2Fadhoc%2Fgrpc_experiment;config=default/log

Any idea? I am not sure if we should tackle the first error first, or the second error (I am not sure if it's right to add `def.hpp`).

This `def.hpp` file comes from this PR https://github.com/protocolbuffers/upb/pull/265, so it sounds like it's a right thing to do to manually include that file.

Taking one of these to dive into:
```
src/core/ext/upb-generated/envoy/api/v2/auth/.libs/cert.o:(.data.rel.local+0x0): multiple definition of `envoy_api_v2_auth_cert_proto_upbdefinit'
src/core/ext/upb-generated/envoy/api/v2/auth/.libs/cert.o:(.data.rel.local+0x0): first defined here
```

`envoy_api_v2_auth_cert_proto_upbdefinit` is only found in these 2 places:
https://github.com/grpc/grpc/blob/master/src/core/ext/upb-generated/envoy/api/v2/auth/cert.upbdefs.c#L42 and https://github.com/grpc/grpc/blob/master/src/core/ext/upb-generated/envoy/api/v2/auth/cert.upbdefs.h#L22, and nobody else is trying to `#include` `src/core/ext/upb-generated/envoy/api/v2/auth/cert.upbdefs.c` directly. So why did it complain `envoy_api_v2_auth_cert_proto_upbdefinit` is defined in multiple places?

---

The PHP extension was built slightly differently in the regular PR tests vs here in the xds tests so maybe that's why this isn't caught when running PR tests in #21941. I can certainly look into adding that. As it stands, the xds tests for PHP is only running on `master`.

**Comment by stanley-cheung** [comment]
I might have gotten past that error by including header files into my PHP extension. Now I got to a different error:

```
libtool: link: `include/grpc/byte_buffer.lo' is not a valid libtool object
Makefile:3583: recipe for target 'grpc.la' failed
make: *** [grpc.la] Error 1
ERROR: `make' failed
```

Still looking into it...

**Comment by stanley-cheung** [comment]
Got past the error above. Now the error becomes:
```
libtool: link: `src/core/ext/filters/client_channel/config_selector.lo' is not a valid libtool object
Makefile:3473: recipe for target 'grpc.la' failed
make: *** [grpc.la] Error 1
ERROR: `make' failed
```

Wut

**Comment by markdroth** [comment]
Nico may be able to help with the link error.

Why was this problem not caught in the tests that ran on the PR?  Do we not build PHP at all in the PR tests?  We should definitely make sure that this kind of problem gets caught there.

**Comment by nicolasnoble** [comment]
> The PHP extension was built slightly differently in the regular PR tests vs here in the xds tests so maybe that's why this isn't caught when running PR tests in #21941. I can certainly look into adding that. As it stands, the xds tests for PHP is only running on `master`.

Is this difficult to address?

**Comment by stanley-cheung** [comment]
> > The PHP extension was built slightly differently in the regular PR tests vs here in the xds tests so maybe that's why this isn't caught when running PR tests in #21941. I can certainly look into adding that. As it stands, the xds tests for PHP is only running on `master`.
> 
> Is this difficult to address?

Ironically, the tests are already there - https://github.com/grpc/grpc/blob/master/src/php/bin/build_all_docker_images.sh and https://github.com/grpc/grpc/blob/master/src/php/bin/run_all_docker_images.sh that were introduced relatively recently to test PHP build against various PHP runtime. They just weren't hooked up to the kokoro tests yet. The concern was that this script wasn't optimized much at all so it's taking a long time to run (an hour or so) so I was hoping to make it better before hooking them up. 

Those tests are building PHP in 9 different docker images, sequentially. This problem would have been caught with the 1st one. So to hook this up to the kokoro tests, I can certainly make a new job for it and only build the 1st image and would have caught a class of problems already. I can certainly do that within a week.

**Comment by markdroth** [comment]
Great, thanks, Stanley!

So now we just need to figure out what's causing this link problem to begin with.  Nico, any suggestions?

**Comment by markdroth** [comment]
I was just talking to Nico about this, and it seems that we do actually have tests that run on every PR that build PHP.  And those tests are passing.  So although I agree that we should make sure the PHP xDS tests run on every PR, I don't think that's actually the root problem here.  I think the root problem is that the PHP xDS tests are building grpc-php using some mechanism that is different from the one in the PR tests, which we know works.

Can we just change the PHP xDS tests to build using the same mechanism as the one we're using in the PR tests?

**Comment by markdroth** [comment]
Note that I think we need to resolve this before we can back-port the upb changes to the 1.30 branch.  We'd like to do that ASAP so that we can release 1.30.1 with the logging changes, which will make it much easier for us to debug problems when users encounter them.

**Comment by stanley-cheung** [comment]
I'll need to double check, but it might need to be the other way round: the current way PHP builds the extension in the xDS test resembles probably the closest to how an end user would have built the PHP extension (i.e. the tgz archive, `pecl install` way).

I do see that way in here: https://github.com/grpc/grpc/blob/master/test/distrib/php/run_distrib_test.sh#L23 but I am not sure that's being run as part of the PR tests.

And then there are these way for the interop tests: https://github.com/grpc/grpc/blob/master/tools/dockerfile/interoptest/grpc_interop_php7/build_interop.sh#L34 and this is _not_ the way how end user would have built the PHP extension. (This way effectively installs c-core separately first.)

In any case, if we were to change the way how the PHP xDS tests build the extension, we would still need to resolve this issue, and I probably still should add that way of building the PHP extension as a new test.

**Comment by stanley-cheung** [comment]
Working on #23321 to add this additional PHP tests to be run per PR.

**Comment by stanley-cheung** [comment]
I had been working with @nicolasnoble and here's to document some of my findings so far.

In summary: The PHP build process has trouble dealing with filenames with multiple dots, like: `src/core/ext/upb-generated/validate/validate.upbdefs.c`. This causes PHP's `./configure` build step to generate a `Makefile` with conflicting recipes.

Details:

Let's take an example. After #21941, we have pairs of files like these 2:
```
    src/core/ext/upb-generated/validate/validate.upb.c
    src/core/ext/upb-generated/validate/validate.upbdefs.c
```

In the PHP build process, after the `./configure` step, a `Makefile` is generated with these 2 recipes:
```
src/core/ext/upb-generated/validate/validate.lo: /home/stanleycheung/grpc-1.31.0dev/src/core/ext/upb-generated/validate/validate.upb.c
        $(LIBTOOL) --mode=compile $(CC) -fvisibility=hidden     -DOPENSSL_NO_ASM -D_GNU_SOURCE -DWIN32_LEAN_AND_MEAN     -D_HAS_EXCEPTIONS=0 -DNOMINMAX -DGRPC_ARES=0     -DGRPC_POSIX_FORK_ALLOW_PTHREAD_ATFORK=1 -I. -I/home/stanleycheung/grpc-1.31.0dev $(COMMON_FLAGS) $(CFLAGS_CLEAN) $(EXTRA_CFLAGS)  -c /home/stanleycheung/grpc-1.31.0dev/src/core/ext/upb-generated/validate/validate.upb.c -o src/core/ext/upb-generated/validate/validate.lo

src/core/ext/upb-generated/validate/validate.lo: /home/stanleycheung/grpc-1.31.0dev/src/core/ext/upb-generated/validate/validate.upbdefs.c
        $(LIBTOOL) --mode=compile $(CC) -fvisibility=hidden     -DOPENSSL_NO_ASM -D_GNU_SOURCE -DWIN32_LEAN_AND_MEAN     -D_HAS_EXCEPTIONS=0 -DNOMINMAX -DGRPC_ARES=0     -DGRPC_POSIX_FORK_ALLOW_PTHREAD_ATFORK=1 -I. -I/home/stanleycheung/grpc-1.31.0dev $(COMMON_FLAGS) $(CFLAGS_CLEAN) $(EXTRA_CFLAGS)  -c /home/stanleycheung/grpc-1.31.0dev/src/core/ext/upb-generated/validate/validate.upbdefs.c -o src/core/ext/upb-generated/validate/validate.lo
```

Notice that both of those files are being instructed to be compiled into `src/core/ext/upb-generated/validate/validate.lo`. This causes those `multiple definition of` errors we saw in the first post of this issue.


So I was diving into how the `./configure` script is generating the `Makefile`. I isolated to this bash script snippet:

```
  for ac_src in src/core/ext/filters/census/grpc_context.cc \
    src/core/ext/filters/client_channel/backend_metric.cc \
    ... some more files ...
    src/core/ext/upb-generated/validate/validate.upb.c \
    src/core/ext/upb-generated/validate/validate.upbdefs.c \
    ... some more files ...
    ; do

      IFS=.
      set $ac_src
      ac_obj=$1

      # then the Makefile object recipe is being constructed by
      ...   $ac_obj.lo
```

It looks like that the combination of `IFS` being set to `.` and then the `set` command setting the filename to positional `$1`, will strip the filename before the _first_ dot. So both of these filenames

```
src/core/ext/upb-generated/validate/validate.upb.c
src/core/ext/upb-generated/validate/validate.upbdefs.c
```

becomes

```
src/core/ext/upb-generated/validate/validate.lo
```

I made a simpler script to reproduce:

```
#!/bin/sh                                                                                                    
for ac_src in src/file1.c src/file2.c src/file1.foo.c \                                                      
  ; do                                                                                                        
  IFS=.                                                                                                      
  set $ac_src                                                                                                
  ac_obj=$1                                                                                                  
  echo $ac_obj.lo                                                                                            
done  
```

This outputs:

```
src/file1.lo                                                                                                  
src/file2.lo                                                                                                  
src/file1.lo 
```

So the trouble is that, in the PHP build process, even the `./configure` script is being generated, from a `config.m4` file. So it isn't easy even to try to manually tweak that file.

**Comment by stanley-cheung** [comment]
I also just realized that there are already tests executing this build path for PHP in the `grpc_distribtests` test suite, which is also only run on `master`: https://fusion.corp.google.com/projectanalysis/history/KOKORO/prod%3Agrpc%2Fcore%2Fmaster%2Flinux%2Fgrpc_distribtests.

This test suite has been broken since #21941 is merged. Looks like some Python distribtests are failing too.

**Comment by stanley-cheung** [comment]
This is fixed by the revert #23524

**Comment by stanley-cheung** [comment]
And the PHP xDS test is back to green!!! https://source.cloud.google.com/results/invocations/260e09d2-63f7-4111-80d7-9fdff695f333/targets

**Comment by markdroth** [comment]
This still needs to be fixed.  We reverted the PR temporarily because this PHP build problem was blocking the release, but we're still trying to get the change in, and this problem needs to be fixed before we can do that.

I have a PR out to roll forward this change in #23533.  Please make sure the PHP build passes with that PR.

**Comment by jtattermusch** [comment]
I removed the "RELEASE BLOCKER" label since this is no longer blocking the release.

**Comment by jtattermusch** [comment]
Is there any chance that the PHP build system issue that's causing this would be fixed by  removing the PHP5 support? https://github.com/grpc/grpc/issues/23645

**Comment by stanley-cheung** [comment]
> Is there any chance that the PHP build system issue that's causing this would be fixed by removing the PHP5 support? #23645

Unfortunately no, this PHP build issue persists to PHP 7. The current Artifact Build test (the one that will fail if #23533 is not resolved) is already built on top of a PHP 7.3 docker image. Remember [this discussion](https://github.com/grpc/grpc/pull/23540#discussion_r456752185)?

**Comment by yashykt** [comment]
Is this still a P1?

**Comment by markdroth** [comment]
No, I don't think this is P1.  Moving to P2.

@stanley-cheung, is this fixed now?  Can you try with #23533 and see if this is still happening?

**Comment by stanley-cheung** [comment]
Closing this as #23533 has been merged and there are no new issues with PHP builds.
